src/rgw/rgw_common.h

```diff
@@ -264,21 +266,58 @@ enum RGWObjCategory {
   RGW_OBJ_CATEGORY_MULTIMETA = 3,
 };
 
+struct s3_err {
+public:
+  struct req_state* s;
+  std::string s3_code;
+  std::string message;
+
+  s3_err() = default;
+  s3_err(struct req_state* s) : s(s) {};
+  s3_err(struct req_state* s, std::string s3_code, std::string message) : s(s), s3_code(s3_code), message(message) {}
+};
+
+struct swift_err {
+public:
+  struct req_state* s;
+  std::string swift_code;
+  std::string message;
+
+  swift_err() = default;
+  swift_err(struct req_state* s) : s(s) {};
+  swift_err(struct req_state* s,std::string swift_code, std::string message) : s(s), swift_code(swift_code), message(message) {}
+};
+
 /** Store error returns for output at a different point in the program */
 struct rgw_err {
-  rgw_err();
-  rgw_err(int http, const std::string &s3);
+  rgw_err(req_state* s);
+  virtual ~rgw_err() { };
   void clear();
   bool is_clear() const;
   bool is_err() const;
   friend std::ostream& operator<<(std::ostream& oss, const rgw_err &err);
-
+  virtual void dump() const;
+  req_state *s;
+  bool is_website_redirect;
   int http_ret;
   int ret;
-  std::string s3_code;
-  std::string message;
+
+  boost::variant<s3_err, swift_err> extra;
+  rgw_err(s3_err err) : extra(std::move(err)) { s = err.s;}
+  rgw_err(swift_err err) : extra(std::move(err)) { s = err.s;}
+
+  rgw_err& operator=(s3_err err) { extra = std::move(err); return *this; }
+  rgw_err& operator=(swift_err err) { extra = std::move(err); return *this; }
+
+  bool set_rgw_err(int err_no);
+  const string& get_err_message() const;
+  void set_err_message(const string& err_msg);
+  const string& get_err_code() const;
+  void set_err_code(const string& err_code);
```

Comment:
i'm not sure that the error `message` and `code` should be part of the `rgw_err` interface - that assumes that all error types, including sts, will have them - and if that is the case, they should be in `rgw_err` instead of the variant types

@mdw-at-linuxbox it would really help if you could share your sts subclass of rgw_err so we know exactly what the requirements are